src/cdomains/apron/elinaImplementation.elina.ml

```diff
@@ -0,0 +1,154 @@
+    Abstract1.meet_tcons_array_with man av tcons_ar
+
+  (* Oct *)
+  type t = Elina_oct.t
+  let to_oct = Elina_oct.Abstract1.to_opt_oct
+  let of_oct = Elina_oct.Abstract1.of_opt_oct
+  let widening_thresholds = Elina_oct.elina_abstract0_opt_oct_widening_thresholds
+  let substitute_texpr_with = substitute_texpr_with_alt
+  let manager_is_oct = Elina_oct.manager_is_opt_oct
+  let manager_to_oct = Elina_oct.manager_to_opt_oct
+  let narrowing = Elina_oct.elina_abstract0_opt_oct_narrowing
+  let manager_alloc = Elina_oct.manager_alloc
+
+  (* Poly *)
+  type pt = Elina_poly.loose Elina_poly.t
+  let manager_alloc_loose = Elina_poly.manager_alloc_loose
+
+  (* Other *)
+  (*
+  Using Abstract1.hash like in the apron interface leads to segmentation faults,
+  so this is used as a workaround
+  *)
+  let hash _ x = Environment.hash (Abstract1.env x)
+  let impl () = "Elina"
+
+
+  (***************)
+  (* Bound Texpr *)
+  (***************)
+
+  (* Scalar to Mpqf *)
+  let stm scal =
+    match scal with
+    | Scalar.Float fl -> Mpqf.of_float fl
+    | Scalar.Mpqf mpqf -> mpqf
+    | Scalar.Mpfrf mpfrf -> Mpfrf.to_mpqf mpfrf
+
+  (* Mpqf to scalar *)
+  (* Creates an interval from two passed Mpqf values *)
+  let mts m1 m2 =
+    Interval.of_scalar (Scalar.of_mpqf m1) (Scalar.of_mpqf m2)
+
+  (* Binop/Mod case *)
+  let bound_texpr_mod i1 _ s1 s2 = 
+    (* There seem to be SO many cases for mod, I dont quite get it *)
+    let inf1 = stm i1 in  
+    let sup1 = stm s1 in
+    let sup2 = stm s2 in
+    let zero = Mpqf.of_int 0 in
+    let one = Mpqf.of_int 1 in
+    let sup2mo = Mpqf.sub sup2 one in
+    (* If dividend is always positive the range is [0;S2-1] or possibly a smaller range *)
+    if Mpqf.sgn inf1 = 1 && Mpqf.sgn sup1 = 1 then
+      mts zero sup2mo
+    (* If dividend is always negative the range is [-S2+1;0] or possibly a smaller range *)
+    else if Mpqf.sgn inf1 = -1 && Mpqf.sgn sup1 = -1 then
+      mts (Mpqf.neg sup2mo) zero
+    (* If dividen could be both negative or positive the range could be [-S2+1;S2-1]*)
+    else
+      mts (Mpqf.neg sup2mo) sup2mo
+    (* This fails to consider many subcases such as the fact that [1;2]%[3;4]
+    would be in range [1;2] not [1;3], etc. Also I'm not even sure I understand the negative
+    result thing correctly*)
+
+  (* Binop/Pow case *)
+  let rec bound_texpr_pow_rec a b = 
+    let one = Mpqf.of_int 1 in
+    if Mpqf.sgn b = 1 then 
+      Mpqf.mul a (bound_texpr_pow_rec a (Mpqf.sub b one))
+    else if Mpqf.sgn b = -1 then
+      Mpqf.div one (bound_texpr_pow_rec a (Mpqf.neg b))
+    else
+      one
+    
+  let bound_texpr_pow i1 i2 s1 s2 =
+    (* TO DO *)
+    mts (bound_texpr_pow_rec (stm i1) (stm i2)) (bound_texpr_pow_rec (stm s1) (stm s2))
+
+  (* Unop/Sqrt case *)
+  let bound_texpr_sqrt inf sup =
+    (* Very poor sqrt approximation *)
+    mts (Mpqf.of_int 0) (Mpqf.add (Mpqf.div (stm sup) (Mpqf.of_int 2)) (Mpqf.of_int 1))
+
+  (* Workaround for Abstract1.bound_texpr *)

```

Comment:
Could you please also reference the issue, if there is one?

Reply:
Added a link to it